src/defrag.c

```diff
@@ -1065,48 +1129,64 @@ void activeDefragCycle(void) {
                 break; /* this will exit the function and we'll continue on the next cycle */
             }
 
+            /* This array of structures holds the parameters for all defragmentation stages. */
+            defragStage defrag_stages[] = {
+                {db->keys, defragScanCallback, &(defragCtx){ db, slot }},
+                {db->expires, scanCallbackCountScanned, NULL},
+                {(current_db == 0 && slot == 0) ? server.pubsub_channels : NULL, defragPubsubScanCallback,
+                    &(defragPubSubCtx){ server.pubsub_channels, getClientPubSubChannels, slot }},
+                {server.pubsubshard_channels, defragPubsubScanCallback,
+                    &(defragPubSubCtx){ server.pubsubshard_channels, getClientPubSubShardChannels, slot }}
+            };
             if (!defrag_later_item_in_progress) {
-                /* Scan the keyspace dict unless we're scanning the expire dict. */
-                if (!expires_cursor)
-                    cursor = kvstoreDictScanDefrag(db->keys, slot, cursor,
-                                                   defragScanCallback,
-                                                   &defragfns, &ctx);
-                /* When done scanning the keyspace dict, we scan the expire dict. */
-                if (!cursor)
-                    expires_cursor = kvstoreDictScanDefrag(db->expires, slot, expires_cursor,
-                                                           scanCallbackCountScanned,
-                                                           &defragfns, NULL);
+                /* Iterates over all defragmentation stages, starting from the last unfinished stage. */
+                int num_stages = sizeof(defrag_stages) / sizeof(defrag_stages[0]);
+                serverAssert(defrag_stage < num_stages);
+                for (int i = defrag_stage; i < num_stages; i++) {
+                    if (defrag_stage != i) break; /* previous stage not finished */
+                    if (defrag_stages[i].kvs) {
+                        defrag_cursor = kvstoreDictScanDefrag(defrag_stages[i].kvs, slot, defrag_cursor,
+                            defrag_stages[i].scanfn, &defragfns, defrag_stages[i].ctx);
+                    }
+
+                    /* Move to next stage if current stage completed */
+                    if (!defrag_cursor) defrag_stage++;
```

Comment:
it's a little bit odd that we choose a slot, then iterate on all the "stages", and then move to the next slot.
it seems more logical to pick a stage, defrag all the slots before moving to the next stage.
am i missing something, or is that what happens?
is there any reason why the other approach won't work or is more complicated?

Reply:
already implemented.